HPXMLtoOpenStudio/resources/waterheater.rb

```diff
@@ -2189,4 +2189,103 @@ def self.get_water_heater_solar_fraction(water_heating_system, hpxml_bldg)
     end
     return solar_fraction.to_f
   end
+
+  # Add an EMS program for calculating unmet showers using sensors on mixed water temperature,
+  # shower volume, and water heater use side outlet temperature schedules.
+  #
+  # @param model [OpenStudio::Model::Model] OpenStudio Model object
+  # @param water_heating_systems [Array<HPXML::WaterHeatingSystem>] The HPXML water heaters of interest
+  # @param plantloop_map [Hash] Map of HPXML System ID => OpenStudio PlantLoop objects
+  # @param mw_temp_schedule [OpenStudio::Model::ScheduleConstant] The mixed water temperature schedule
+  # @param showers_schedule_name [String] The name of the shower schedule
+  # @return [nil]
+  def self.unmet_showers_program(model, water_heating_systems, plantloop_map, mw_temp_schedule, showers_schedule_name)
+    return if water_heating_systems.empty? || mw_temp_schedule.nil?
+
+    # EMS sensors
+    mixed_setpoint_sensor = Model.add_ems_sensor(
+      model,
+      name: 'res_shower_mixsp',
+      output_var_or_meter_name: 'Schedule Value',
+      key_name: mw_temp_schedule.name
+    )
+
+    shower_flow_sensor = Model.add_ems_sensor(
+      model,
+      name: 'shower_volume',
+      output_var_or_meter_name: 'Schedule Value',
+      key_name: showers_schedule_name
+    )
+
+    # EMS program
+    program = Model.add_ems_program(
+      model,
+      name: 'unmet showers program'
+    )
+    program.addLine('Set ShowerUnmetTime = 0') # Init
+    program.addLine('Set ShowerTime = 0') # Init
+    program.addLine('If WarmupFlag == 0') # Prevent unmet hours in the first hour because of the warmup period
+
+    water_heating_systems.each do |water_heating_system|
+      # Get the water storage tanks for the outlet temperature sensor
+      tank = nil
+      hw_plant_loop = plantloop_map[water_heating_system.id]
+      hw_plant_loop.components.each do |component|
+        if component.to_WaterHeaterMixed.is_initialized
+          tank = component.to_WaterHeaterMixed.get
+        elsif component.to_WaterHeaterStratified.is_initialized
+          object = component.to_WaterHeaterStratified.get
+
+          object_type = object.additionalProperties.getFeatureAsString('ObjectType')
+          next if object_type == Constants::ObjectTypeSolarHotWater # Exclude solar storage from 2 tank systems
+
+          tank = object
+        end
+      end
+
+      wh_temp_sensor = Model.add_ems_sensor(
+        model,
+        name: "#{tank.name} Outlet Temperature",
+        output_var_or_meter_name: 'Water Heater Use Side Outlet Temperature',
+        key_name: tank.name
+      )
+
+      program.addLine("If (#{shower_flow_sensor.name} > 0) && (#{wh_temp_sensor.name} < #{mixed_setpoint_sensor.name})")
```

Comment:
After discussing with @jmaguire1... Since we no longer calculate an unmet *load*, we now only care whether there is a shower for the given timestep. To prevent confusion about how the shower schedule is used (and prevent the need for error-checking if the shower schedule does not correctly jive with the overall fixtures schedule), I'm going to change the shower schedule to be a 0/1 schedule that simply represents the presence of a shower rather than the shower usage.

Reply:
I'm second-guessing this. If the detailed schedule is hourly, you'd like to use a fraction for how much of the hour there were showers. Also related: https://github.com/NREL/OpenStudio-HPXML/issues/1890

Reply:
I also wonder if the `ShowerUnmetTime` and `ShowerTime` variables below should be multiplied by the schedule value, where the schedule value represents how much of the timestep is for showers (stochastic) or the likelihood of showers for a given timestep (non-stochastic).